<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Genomic Coordinate data</title>
    <style>
    * {
        font-family: Calibri;
        }
    </style>
</head>
<body>
<!--Header for the web page-->
<h1 style="color: navy;">Variants associated with Type 1 Diabetes found in genomic coordinate range</h1>

<!--Display table of information for rsIDs found in the input genomic region range-->
{% for table in tables %}
        {{ titles[loop.index] }}
        {{ table|safe }}
    </table>
{% endfor %}


<!--Display information related to the the risk allele frequency columns of the output table-->
<p style="color: maroon; font-family: Calibri;">FIN_freq - risk allele frequency for the Finish population from the 1000 genomes project (a coutry with high Type 1 Diabetes rates)</p>
<p style="color: maroon; font-family: Calibri;">JPT_freq - risk allele frequency for the Japanese population from the 1000 genomes project ((a coutry with average Type 1 Diabetes rates)</p>
<p style="color: maroon; font-family: Calibri;">NGA_freq - risk allele frequency for the Nigerian population from the 1000 genomes project (a coutry with low Type 1 Diabetes rates)</p>
                                                                                                                                            
<!--Display Mahattan plot-->                                                                                                                 
<p><img class="plot" src="data:image/png;base64, {{plot_url}}"></p>
                                                                                                                                            
<!--Create URL for linkage disequilibrium rsID input page-->
<p style="font-family: Calibri"><a href="{{ url_for('R2_index')}}">Linkage disequilibrium calculator</a></p>
</body>
</html>
